Kubernetes Integration Example (#34)
Bodo-inc/Bodo-examples: Kubernetes/README.md

# Bodo Kubernetes Example

This folder contains information about how to deploy a Bodo application with Kubernetes. We deploy Bodo with the [Kubeflow MPI-Operator](https://github.com/kubeflow/mpi-operator), which enables resiliency in the case of Node Failure for long running Bodo applications. 

## Prerequisites

> - <strong>Docker image </strong>: Create a docker image that contains your intended Bodo version and Python scripts. For pi calculation, you may use [bodoaidocker/pi:latest](https://hub.docker.com/r/bodoaidocker/pi/tags). 

Author: Do you have any name preferences?

> - <strong>Docker image </strong>: Create a docker image that contains your intended Bodo version and Python scripts. For pi calculation, you may use [bodoaidocker/pi:latest](https://hub.docker.com/r/bodoaidocker/pi/tags). 

Reviewer: Maybe something like `bodoaidocker/kube-mpioperator-minimal` or `bodoaidocker/k8s-mpioperator-minimal`?

Also maybe we should say that they can use this as a base image for their own docker images and that a pi calculation example is included for testing purposes.

- <strong>AWS </strong>: Have Access-Key/Secret-Key pair.

## Setup

### Step 1: Create a Kubernetes Cluster in EKS

- Install KOPS on your local machine:

```
# Mac
brew install kops
# Linux
curl -LO https://github.com/kubernetes/kops/releases/download/$(curl -s https://api.github.com/repos/kubernetes/kops/releases/latest | grep tag_name | cut -d '"' -f 4)/kops-linux-amd64
chmod +x kops-linux-amd64
sudo mv kops-linux-amd64 /usr/local/bin/kops
```

- Create a cluster:

To change the number of instances, modify the `node-count` argument and to change the worker nodes update `node-size`. `master-size` refers to the leader that manages K8s but doesn’t do any computation, so you should keep the instance small. If you are in a different region, change `zones` argument. 

```
export KOPS_CLUSTER_NAME=imesh.k8s.local

> export KOPS_STATE_STORE=s3://mpijob-cluster-bucket

Author: - I did not make any bucket. Why is that needed?
- I'll change the name to a placeholder.

```

You can check whether the MPI Job custom resource is installed via:

```
kubectl get crd
```

The output should include mpijobs.kubeflow.org like the following:

```
NAME                   CREATED AT
mpijobs.kubeflow.org   2022-01-03T21:19:10Z
```

### Step 3: Run your Bodo application

- You may use the `example-mpijob.yaml` as it is for the sake of this example. If you wish to modify it with your experiment configuration; update `spec.slotsPerWorker` with the number of physical cores (not vcpus) on each Node and set `spec.mpiReplicaSpecs.Worker.replicas` with the number of workers node in your cluster. At last, make sure `-n` matches your cluster.

- Run the example by deploying it over your cluster with `kubectl create -f example-mpijob.yaml`. This should add 1 pod to each worker and a launcher pod to your master node. 

- View the generated pods by this deployment with `kubectl get pods`. You may inspect any logs by looking at the individual pod's logs.

### Step 4: Get the Results

- When the job finishes running, your launcher pod will change its status to completed and any stdout information can be found in the logs of the launcher pod:

```

> PODNAME=$(kubectl get pods -o=name)

Author: I tried to use this `PODNAME=$(kubectl get pods -l mpi_job_name=bodo-example,mpi_role_type=launcher -o name)` but it did not work. The above was the only thing that worked.

> PODNAME=$(kubectl get pods -o=name)

Reviewer: Sounds good, thanks.

# Bodo Kubernetes Example

> This folder contains information about how to deploy a Bodo application with Kubernetes. We deploy Bodo using the [Kubeflow MPI-Operator](https://github.com/kubeflow/mpi-operator), which enables running MPI applications in a Kubernetes environment. In typical Kubernetes fashion, this also provides resiliency in the case of Node Failure for long running Bodo applications. 

Reviewer: ```suggestion
Bodo workloads can be deployed with Kubernetes using the [Kubeflow MPI-Operator](https://github.com/kubeflow/mpi-operator), which enables running MPI applications in a Kubernetes environment. In typical Kubernetes fashion, this also provides resiliency in the case of Node Failure for long running Bodo applications. 
```

## Prerequisites

- Access to a Kubernetes cluster such as AWS EKS.

> - Create a docker image that contains your intended Bodo version and Python scripts. A [pi calculation docker image](https://hub.docker.com/r/bodoaidocker/kube-mpioperator-minimal/tags) is built as an example for testing purposes. To create your own docker images, use the example docker file and python script available at [docker folder](https://github.com/Bodo-inc/Bodo-examples/tree/master/Kubernetes/docker) as the base. 

Reviewer: ```suggestion
- Create a docker image that contains your intended Bodo version and Python scripts. For reference, see this [Dockerfile](docker/Dockerfile). 
A docker image created from this Dockerfile is also available on DockerHub: [bodoaidocker/kube-mpioperator-minimal](https://hub.docker.com/r/bodoaidocker/kube-mpioperator-minimal/tags). 
You can use this as the base image for your own docker image. 
For testing and validation purposes this image also includes the [pi calculation example](docker/pi.py), which is used in this tutorial.
```

## Setup

> In the following example, we use KOPS to setup Kubernetes Cluster in EKS while this works for any K8s cluster. 

Reviewer: ```suggestion
Bodo can be deployed in any Kubernetes cluster. For the purposes of this example, we set up a Kubernetes Cluster in EKS using KOPS:
```

### Step 1: Create a Kubernetes Cluster in EKS

- Install KOPS on your local machine:

```
# Mac
brew install kops
# Linux
curl -LO https://github.com/kubernetes/kops/releases/download/$(curl -s https://api.github.com/repos/kubernetes/kops/releases/latest | grep tag_name | cut -d '"' -f 4)/kops-linux-amd64
chmod +x kops-linux-amd64
sudo mv kops-linux-amd64 /usr/local/bin/kops
```

- Create a cluster:

Begin by making a bucket in the S3 to use as your `KOPS_STATE_STORE`.  

```
export KOPS_CLUSTER_NAME=imesh.k8s.local
export KOPS_STATE_STORE=s3://<your S3 bucket name>
```

> - Attempt to create your cluster. This creates a cluster of 2 nodes each with 4 cores. To change the number of instances, modify the `node-count` argument and to change the worker nodes update `node-size`. `master-size` refers to the leader that manages K8s but doesn’t do any computation, so you should keep the instance small. If you are in a different region, change `zones` argument. 

Reviewer: ```suggestion
- Attempt to create your cluster. This creates a cluster of 2 nodes each with 4 cores. To change the number of instances, modify the `node-count` argument and to change the worker nodes update `node-size`. `master-size` refers to the leader that manages K8s but doesn’t do any computation, so you should keep the instance small. You can deploy the cluster in a different AWS region and availability zone by modifying the `zones` argument. 
```

```
kops create cluster \
--node-count=2 \
--node-size=c5.2xlarge \
--master-size=c5.large \
--zones=us-east-2c \
--name=${KOPS_CLUSTER_NAME}
```

- Finish creating the cluster with the below command. This might take several minutes to finish:

```
kops update cluster --name $KOPS_CLUSTER_NAME --yes --admin
```
- Verify the cluster setup is finished by:

```
kops validate cluster
```

### Step 2: Install MPIJob

- The most up-to-date installation guide is available at [MPI-Operator Github](https://github.com/kubeflow/mpi-operator). You may also follow the below steps :

```
git clone https://github.com/kubeflow/mpi-operator
cd mpi-operator
kubectl apply -f deploy/v2beta1/mpi-operator.yaml
```

You can check whether the MPI Job custom resource is installed via:

```
kubectl get crd
```

The output should include mpijobs.kubeflow.org like the following:

```
NAME                   CREATED AT
mpijobs.kubeflow.org   2022-01-03T21:19:10Z
```

### Step 3: Run your Bodo application

> - You may use the `example-mpijob.yaml` as it is for the sake of this example. If you wish to modify it with your experiment configuration; update `spec.slotsPerWorker` with the number of physical cores (not vcpus) on each Node and set `spec.mpiReplicaSpecs.Worker.replicas` with the number of worker nodes in your cluster. At last, make sure `-n` is equal to `spec.mpiReplicaSpecs.Worker.replicas` multiplied by `spec.slotsPerWorker`.

Reviewer: ```suggestion
- Define a kubernetes resource for your Bodo workload, such as the one defined in [`example-mpijob.yaml`](example-mpijob.yaml) that runs the [pi calculation example](docker/pi.py). You can modify it based on your cluster configuration: update `spec.slotsPerWorker` with the number of physical cores (_not_ vCPUs) on each node and set `spec.mpiReplicaSpecs.Worker.replicas` to the number of worker nodes in your cluster. Lastly, make sure `-n` is equal to `spec.mpiReplicaSpecs.Worker.replicas` multiplied by `spec.slotsPerWorker`, i.e. the total number of physical cores on your worker nodes.
```

> - Run the example by deploying it over your cluster with `kubectl create -f example-mpijob.yaml`. This should add 1 pod to each worker and a launcher pod to your master node. 

Reviewer: ```suggestion
- Run the example by deploying it in your cluster with `kubectl create -f example-mpijob.yaml`. This should add 1 pod to each worker and a launcher pod to your master node. 
```

- View the generated pods by this deployment with `kubectl get pods`. You may inspect any logs by looking at the individual pod's logs.

### Step 4: Get the Results

- When the job finishes running, your launcher pod will change its status to completed and any stdout information can be found in the logs of the launcher pod:

```
PODNAME=$(kubectl get pods -o=name)
kubectl logs -f ${PODNAME}

```

## Teardown

> - When a job has finished running, you can remove it by running `kubectl delete -f example-mpijob.yaml`. If you want to delete the MPI-Operator crd, please follow any steps on the [MPI-Operator Github](https://github.com/kubeflow/mpi-operator).

Reviewer: ```suggestion
- When a job has finished running, you can remove it by running `kubectl delete -f example-mpijob.yaml`. If you want to delete the MPI-Operator crd, please follow the steps on the [MPI-Operator Github repository](https://github.com/kubeflow/mpi-operator).
```

# Running a Bodo workload in Kubernetes

Bodo workloads can be deployed with Kubernetes using the [Kubeflow MPI-Operator](https://github.com/kubeflow/mpi-operator) which enables running MPI applications in a Kubernetes environment. In typical Kubernetes fashion, this also provides resiliency in the case of Node Failure for long running Bodo applications. 

## Prerequisites

- Access to a Kubernetes cluster such as AWS EKS.

> - Create a docker image that contains your intended Bodo version and Python scripts. For reference, see this [Dockerfile](docker/Dockerfile). 

Reviewer: I agree, let's add a sentence about needing to upload the image to some kind of registry like DockerHub so K8s can pull it. In case of private registries they can follow instructions like https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/ .

A docker image created from this Dockerfile is also available on DockerHub: [bodoaidocker/kube-mpioperator-minimal](https://hub.docker.com/r/bodoaidocker/kube-mpioperator-minimal/tags). 
You can use this as the base image for your own docker image. 
For testing and validation purposes this image also includes the [pi calculation example](docker/pi.py), which is used in this tutorial.

## Setup
Bodo can be deployed in any Kubernetes cluster. For the purposes of this example, we set up a Kubernetes Cluster in EKS using KOPS:

### Step 1: Create a Kubernetes Cluster in EKS

- Install KOPS on your local machine:

```
# Mac
brew install kops
# Linux
curl -LO https://github.com/kubernetes/kops/releases/download/$(curl -s https://api.github.com/repos/kubernetes/kops/releases/latest | grep tag_name | cut -d '"' -f 4)/kops-linux-amd64
chmod +x kops-linux-amd64
sudo mv kops-linux-amd64 /usr/local/bin/kops
```

- Create a cluster:

Begin by making a bucket in the S3 to use as your `KOPS_STATE_STORE`.  

```
export KOPS_CLUSTER_NAME=imesh.k8s.local
export KOPS_STATE_STORE=s3://<your S3 bucket name>
```

> - Attempt to create your cluster. This creates a cluster of 2 nodes each with 4 cores. To change the number of instances, modify the `node-count` argument and to change the worker nodes update `node-size`. `master-size` refers to the leader that manages K8s but doesn’t do any computation, so you should keep the instance small. You can deploy the cluster in a different AWS region and availability zone by modifying the `zones` argument. 

Author: Hi Nick, would you please elaborate more on this? it already says "the leader only manages K8s but doesn’t do any computation". How this should change to become more clarified? thanks

> - Attempt to create your cluster. This creates a cluster of 2 nodes each with 4 cores. To change the number of instances, modify the `node-count` argument and to change the worker nodes update `node-size`. `master-size` refers to the leader that manages K8s but doesn’t do any computation, so you should keep the instance small. You can deploy the cluster in a different AWS region and availability zone by modifying the `zones` argument. 

Reviewer 2: Its a bit of a nitpick but I would say "the leader only manages K8s but doesn’t do any Bodo computation". Managing the cluster may require computation for complex configurations, so I want to avoid advising on picking the KOPs leader in general.

```
kops create cluster \
--node-count=2 \
--node-size=c5.2xlarge \
--master-size=c5.large \
--zones=us-east-2c \
--name=${KOPS_CLUSTER_NAME}
```

- Finish creating the cluster with the below command. This might take several minutes to finish:

```
kops update cluster --name $KOPS_CLUSTER_NAME --yes --admin
```
- Verify the cluster setup is finished by:

```
kops validate cluster
```

### Step 2: Install MPIJob

- The most up-to-date installation guide is available at [MPI-Operator Github](https://github.com/kubeflow/mpi-operator). You may also follow the below steps :

```

> git clone https://github.com/kubeflow/mpi-operator

Reviewer: Yeah I tested it on their master. But I agree, might be worth locking it to their most recent release.

> git clone https://github.com/kubeflow/mpi-operator

Author: @sahil1105 should we remove this part and just refer to the `mpi-operator` repo instead?

> git clone https://github.com/kubeflow/mpi-operator

Reviewer: I think we could do something like `git clone https://github.com/kubeflow/mpi-operator --branch <VERIFIED_RELEASE_TAG>` in our instructions and similarly, in the link we could update it to be https://github.com/kubeflow/mpi-operator/tree/<VERIFIED_RELEASE_TAG>

```
NAME                   CREATED AT
mpijobs.kubeflow.org   2022-01-03T21:19:10Z
```

### Step 3: Run your Bodo application

- Define a kubernetes resource for your Bodo workload, such as the one defined in [`example-mpijob.yaml`](example-mpijob.yaml) that runs the [pi calculation example](docker/pi.py). You can modify it based on your cluster configuration: update `spec.slotsPerWorker` with the number of physical cores (_not_ vCPUs) on each node and set `spec.mpiReplicaSpecs.Worker.replicas` to the number of worker nodes in your cluster. Lastly, make sure `-n` is equal to `spec.mpiReplicaSpecs.Worker.replicas` multiplied by `spec.slotsPerWorker`, i.e. the total number of physical cores on your worker nodes. If you're using the cluster configuration as defined in step 1, you do not need to modify anything.

- Run the example by deploying it in your cluster with `kubectl create -f example-mpijob.yaml`. This should add 1 pod to each worker and a launcher pod to your master node. 

- View the generated pods by this deployment with `kubectl get pods`. You may inspect any logs by looking at the individual pod's logs.

### Step 4: Get the Results

> - When the job finishes running, your launcher pod will change its status to completed and any stdout information can be found in the logs of the launcher pod:

Reviewer: The launcher pod has the logs from the computation and remains available until the resource is removed, right?

> - When the job finishes running, your launcher pod will change its status to completed and any stdout information can be found in the logs of the launcher pod:

Author: yes, the launcher keeps the logs

> - When the job finishes running, your launcher pod will change its status to completed and any stdout information can be found in the logs of the launcher pod:

Reviewer: @njriasan Does that address it or were you referring to something else?

> - When the job finishes running, your launcher pod will change its status to completed and any stdout information can be found in the logs of the launcher pod:

Reviewer 2: The launcher keeps the logs if it succeeds, but what about if it fails and restarts. Its possible they made it perform better, but in the past this has caused me problems. 

Consider the following test:

1. Run an example on > 8 cores without a license. This will fail running the script because of the license check.
2. Set the number of retries to be 2.

Will the logs contain both errors or just the last error? If we only see 1 error this could be a problem if the error is consistent but may take a long time to occur (i.e. OOM). If a user notices a job is taking multiple retries they shouldn't have to wait for the last one to finish to flush the logs.

> - When the job finishes running, your launcher pod will change its status to completed and any stdout information can be found in the logs of the launcher pod:

Reviewer 2: On MacOS with I `cat` the license I always get an extra `%` character that is garbage and depending on where you open the license there could be extra whitespace breaking it.

Not sure if any of this is helpful, just wanted to share what I did.

> - When the job finishes running, your launcher pod will change its status to completed and any stdout information can be found in the logs of the launcher pod:

Author: > Also for getting the license to work what did you try. Copying it into the env of the yaml file worked for me. https://github.com/Bodo-inc/Bodo-Kubernetes-Example/blob/main/example-mpijob.yaml#L77

@njriasan , I tried your method but it did not work. The logs were weird and not related to the License. Do you have the bandwidth to try it?
Here is the example of the logs that I was getting:

```
:: initializing oneAPI environment ...
  entrypoint.sh: BASH_VERSION = 5.0.3(1)-release
:: mpi -- latest
:: oneAPI environment initialized ::
 
Couldn't resolve pi-launcher... Retrying
Couldn't resolve pi-launcher... Retrying
Couldn't resolve pi-launcher... Retrying
Couldn't resolve pi-launcher... Retrying
Couldn't resolve pi-launcher... Retrying
Couldn't resolve pi-launcher... Retrying
Couldn't resolve pi-launcher... Retrying
Couldn't resolve pi-launcher... Retrying
Couldn't resolve pi-launcher... Retrying
Resolved pi-launcher
Couldn't resolve pi-worker-0.pi-worker... Retrying
Couldn't resolve pi-worker-0.pi-worker... Retrying
Couldn't resolve pi-worker-0.pi-worker... Retrying
Couldn't resolve pi-worker-0.pi-worker... Retrying
Couldn't resolve pi-worker-0.pi-worker... Retrying
Couldn't resolve pi-worker-0.pi-worker... Retrying
Couldn't resolve pi-worker-0.pi-worker... Retrying
Couldn't resolve pi-worker-0.pi-worker... Retrying
Couldn't resolve pi-worker-0.pi-worker... Retrying
Couldn't resolve pi-worker-0.pi-worker... Retrying
Couldn't resolve pi-worker-0.pi-worker
Couldn't resolve pi-worker-1.pi-worker... Retrying
Couldn't resolve pi-worker-1.pi-worker... Retrying
Couldn't resolve pi-worker-1.pi-worker... Retrying
Couldn't resolve pi-worker-1.pi-worker... Retrying
Couldn't resolve pi-worker-1.pi-worker... Retrying
Couldn't resolve pi-worker-1.pi-worker... Retrying
Couldn't resolve pi-worker-1.pi-worker... Retrying
Couldn't resolve pi-worker-1.pi-worker... Retrying
Couldn't resolve pi-worker-1.pi-worker... Retrying
Couldn't resolve pi-worker-1.pi-worker... Retrying
Couldn't resolve pi-worker-1.pi-worker
[mpiexec@pi-launcher] check_exit_codes (../../../../../src/pm/i_hydra/libhydra/demux/hydra_demux_poll.c:117): unable to run bstrap_proxy on pi-worker-0.pi-worker (pid 330, exit code 65280)
[mpiexec@pi-launcher] poll_for_event (../../../../../src/pm/i_hydra/libhydra/demux/hydra_demux_poll.c:159): check exit codes error
[mpiexec@pi-launcher] HYD_dmx_poll_wait_for_proxy_event (../../../../../src/pm/i_hydra/libhydra/demux/hydra_demux_poll.c:212): poll for event error
[mpiexec@pi-launcher] HYD_bstrap_setup (../../../../../src/pm/i_hydra/libhydra/bstrap/src/intel/i_hydra_bstrap.c:1062): error waiting for event
[mpiexec@pi-launcher] HYD_print_bstrap_setup_error_message (../../../../../src/pm/i_hydra/mpiexec/intel/i_mpiexec.c:1014): error setting up the bootstrap proxies
[mpiexec@pi-launcher] Possible reasons:
[mpiexec@pi-launcher] 1. Host is unavailable. Please check that all hosts are available.
[mpiexec@pi-launcher] 2. Cannot launch hydra_bstrap_proxy or it crashed on one of the hosts. Make sure hydra_bstrap_proxy is available on all hosts and it has right permissions.
[mpiexec@pi-launcher] 3. Firewall refused connection. Check that enough ports are allowed in the firewall and specify them with the I_MPI_PORT_RANGE variable.
[mpiexec@pi-launcher] 4. Ssh bootstrap cannot launch processes on remote host. Make sure that passwordless ssh connection is established across compute hosts.
[mpiexec@pi-launcher]  You may try using -bootstrap option to select alternative launcher.```

> - When the job finishes running, your launcher pod will change its status to completed and any stdout information can be found in the logs of the launcher pod:

Author: > The launcher keeps the logs if it succeeds, but what about if it fails and restarts. Its possible they made it perform better, but in the past this has caused me problems.
> 
> Consider the following test:
> 
> 1. Run an example on > 8 cores without a license. This will fail running the script because of the license check.
> 2. Set the number of retries to be 2.
> 
> Will the logs contain both errors or just the last error? If we only see 1 error this could be a problem if the error is consistent but may take a long time to occur (i.e. OOM). If a user notices a job is taking multiple retries they shouldn't have to wait for the last one to finish to flush the logs.

I'll try this and let you know

> - When the job finishes running, your launcher pod will change its status to completed and any stdout information can be found in the logs of the launcher pod:

Reviewer: Let's merge this PR and iterate on this example to resolve these issue in the next PR.
@nmshafie1993 Let's capture these in JIRA issues so we have the details handy, thanks!

> - When the job finishes running, your launcher pod will change its status to completed and any stdout information can be found in the logs of the launcher pod:

Author: Sure, I'll add Nick's comments to the Resiliency blog Jira task.

# Running a Bodo workload in Kubernetes

Bodo workloads can be deployed with Kubernetes using the [Kubeflow MPI-Operator](https://github.com/kubeflow/mpi-operator) which enables running MPI applications in a Kubernetes environment. In typical Kubernetes fashion, this also provides increased resiliency in the case of Node Failure for long running Bodo applications. 

## Prerequisites

- Access to a Kubernetes cluster such as AWS EKS.

> - Create a docker image that contains your intended Bodo version and Python scripts and upload it to a docker registry such as Docker Hub so that K8s can pull it. For reference, see this [Dockerfile](docker/Dockerfile).
> In case of private registries, follow instructions from [here](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/).
> A docker image created from this Dockerfile is also available on DockerHub: [bodoaidocker/kube-mpioperator-minimal](https://hub.docker.com/r/bodoaidocker/kube-mpioperator-minimal/tags). 
> You can use this as the base image for your own docker image. For testing and validation purposes this image also includes the [pi calculation example](docker/pi.py), which is used in this tutorial.

Reviewer: ```suggestion
- Create a docker image that contains your intended Bodo version and Python scripts and upload it to a docker registry such as Docker Hub so that K8s can pull it.
For reference, see this [Dockerfile](docker/Dockerfile).
A docker image created from this Dockerfile is also available on DockerHub: [bodoaidocker/kube-mpioperator-minimal](https://hub.docker.com/r/bodoaidocker/kube-mpioperator-minimal/tags).
You can use this as the base image for your own docker image. For testing and validation purposes this image also includes the [pi calculation example](docker/pi.py), which is used in this tutorial.
In case of private registries, follow instructions from [here](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/).
```
Just moved it to the end.

chmod +x kops-linux-amd64
sudo mv kops-linux-amd64 /usr/local/bin/kops
```

- Create a cluster:

Begin by making a bucket in the S3 to use as your `KOPS_STATE_STORE`.  

```
export KOPS_CLUSTER_NAME=imesh.k8s.local
export KOPS_STATE_STORE=s3://<your S3 bucket name>
```

- Attempt to create your cluster: 

This creates a cluster of 2 nodes each with 4 cores. To change the number of instances, modify the `node-count` argument and to change the worker nodes update `node-size`. `master-size` refers to the leader that manages K8s but doesn’t do any computation, so you should keep the instance small. You can deploy the cluster in a different AWS region and availability zone by modifying the `zones` argument. 

```
kops create cluster \
--node-count=2 \
--node-size=c5.2xlarge \
--master-size=c5.large \
--zones=us-east-2c \
--name=${KOPS_CLUSTER_NAME}
```

- Finish creating the cluster with the below command. This might take several minutes to finish:

```
kops update cluster --name $KOPS_CLUSTER_NAME --yes --admin
```
- Verify the cluster setup is finished by:

```
kops validate cluster
```

### Step 2: Install MPIJob

> - The most up-to-date installation guide is available at [MPI-Operator Github](https://github.com/kubeflow/mpi-operator). You may also get the [latest verified release tag](https://github.com/kubeflow/mpi-operator/releases/latest) and follow the below steps :

Reviewer: ```suggestion
- The most up-to-date installation guide is available at [MPI-Operator Github](https://github.com/kubeflow/mpi-operator). This example was tested using [v0.3.0](https://github.com/kubeflow/mpi-operator/tree/v0.3.0), as shown below:
```

```

> git clone https://github.com/kubeflow/mpi-operator --branch <VERIFIED_RELEASE_TAG>

Reviewer: We need to insert a value for `<VERIFIED_RELEASE_TAG>` :)
Let's confirm that their latest release (v0.3.0) works, and then hard-code the value.

Reviewer: ```suggestion
git clone https://github.com/kubeflow/mpi-operator --branch v0.3.0
```
